dozer-recordstore/src/lib.rs

```diff
@@ -0,0 +1,542 @@
+            let (ptr, value) = read(ptr);
+            (ptr, Field::Point(value))
+        }
+        FieldType::Duration => {
+            let (ptr, value) = read(ptr);
+            (ptr, Field::Duration(value))
+        }
+    }
+}
+
+fn size(fields: &[Option<FieldType>]) -> usize {
+    fields
+        .iter()
+        .copied()
+        // Filter out None -> NULL -> size 0
+        .flatten()
+        .map(|field| match field {
+            FieldType::UInt => aligned_size_of::<u64>(),
+            FieldType::U128 => aligned_size_of::<u128>(),
+            FieldType::Int => aligned_size_of::<i64>(),
+            FieldType::I128 => aligned_size_of::<i128>(),
+            FieldType::Float => aligned_size_of::<OrderedFloat<f64>>(),
+            FieldType::Boolean => aligned_size_of::<bool>(),
+            FieldType::String => aligned_size_of::<String>(),
+            FieldType::Text => aligned_size_of::<String>(),
+            FieldType::Binary => aligned_size_of::<Vec<u8>>(),
+            FieldType::Decimal => aligned_size_of::<Decimal>(),
+            FieldType::Timestamp => aligned_size_of::<DateTime<FixedOffset>>(),
+            FieldType::Date => aligned_size_of::<NaiveDate>(),
+            FieldType::Json => aligned_size_of::<JsonValue>(),
+            FieldType::Point => aligned_size_of::<DozerPoint>(),
+            FieldType::Duration => aligned_size_of::<DozerDuration>(),
+        })
+        .sum()
+}
+
+#[derive(Hash, Serialize, Deserialize, Debug, PartialEq, Eq)]
+#[serde(crate = "dozer_types::serde")]
+pub enum FieldRef<'a> {
+    UInt(u64),
+    U128(u128),
+    Int(i64),
+    I128(i128),
+    Float(OrderedFloat<f64>),
+    Boolean(bool),
+    String(&'a str),
+    Text(&'a str),
+    Binary(&'a [u8]),
+    Decimal(Decimal),
+    Timestamp(DateTime<FixedOffset>),
+    Date(NaiveDate),
+    Json(JsonValue),
+    Point(DozerPoint),
+    Duration(DozerDuration),
+    Null,
+}
+
+impl FieldRef<'_> {
+    pub fn cloned(&self) -> Field {
+        match self {
+            FieldRef::UInt(v) => Field::UInt(*v),
+            FieldRef::U128(v) => Field::U128(*v),
+            FieldRef::Int(v) => Field::Int(*v),
+            FieldRef::I128(v) => Field::I128(*v),
+            FieldRef::Float(v) => Field::Float(*v),
+            FieldRef::Boolean(v) => Field::Boolean(*v),
+            FieldRef::String(v) => Field::String((*v).to_owned()),
+            FieldRef::Text(v) => Field::Text((*v).to_owned()),
+            FieldRef::Binary(v) => Field::Binary((*v).to_vec()),
+            FieldRef::Decimal(v) => Field::Decimal(*v),
+            FieldRef::Timestamp(v) => Field::Timestamp(*v),
+            FieldRef::Date(v) => Field::Date(*v),
+            FieldRef::Json(v) => Field::Json(v.clone()),
+            FieldRef::Point(v) => Field::Point(*v),
+            FieldRef::Duration(v) => Field::Duration(*v),
+            FieldRef::Null => Field::Null,
+        }
+    }
+}
+
+impl RecordRef {
+    pub fn new(fields: &[Field]) -> Self {
```

Comment:
Should we take ownership of `fields` because it's moved into the inner representation?